File: chap07_calcium/intercell_cal_wave.py

#   -------------------------------------------------------------------
# 
#    Use the method of lines to compute an intercellular calcium wave. This
#    version uses the modal model of the IP3 receptor, but this can be
#    changed. Here, IP3 is also diffusing and has simple reaction terms.
# 
#    This version does not have jump conditions at the intercellular boundaries, 
#    as Python doesn't seem to like DAEs. So the jump conditions have been 
#    converted to additional ODEs instead.
# 
#    For Chapter 7, Section 7.10.2 of
#    Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
#  
#    Written by James Keener and James Sneyd.
#  
#   ------------------------------------------------------------------- 


import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp


# The right-hand side for PDE (MoL) simulation:
def pde_rhs(t, s):
    # There are four variables:  c, ce, h, IP
    # only c, ce, and IP are diffusing

    c = s[: N]  # calcium
    ce = s[N : 2 * N]  # ER calcium
    h = s[2 * N : 3 * N]  # h
    IP = s[3 * N :]  # IP3

    lam_c = Dc / delx ** 2
    lam_e = De / delx ** 2
    lam_p = DIP / delx ** 2

    out = coscrhs(t, s)
    sc = np.concatenate([np.array([1]), 2 * np.ones(N - 2), np.array([1])])

    Fc = (
        lam_c * (-sc * c + np.concatenate([np.array([0]), c[:-1]]) + np.concatenate([c[1:], np.array([0])]))
        + out[: N]
    )
    Fce = (
        lam_e
        * (-sc * ce + np.concatenate([np.array([0]), ce[:-1]]) + np.concatenate([ce[1:], np.array([0])]))
        + out[N : 2 * N]
    )
    Fh = out[2 * N : 3 * N]  # reaction only, no diffusion
    FIP = (
        lam_p
        * (-sc * IP + np.concatenate([np.array([0]), IP[:-1]]) + np.concatenate([IP[1:], np.array([0])]))
        + out[3 * N :]
    )

    # now put in the intercellular boundaries. Differentiate the jump conditions
    # to get these ODEs for the boundary points.
    for j in range(Ncell - 1):
        d = int((j + 1) * nc)

        Fc[d] = (Fc[d-1]*(1+muc) + muc*Fc[d+2])/(2*muc+1)
        Fc[d + 1] = (Fc[d+2]*(1+muc) + muc*Fc[d-1])/(2*muc+1)
        
        Fce[d] = (Fce[d-1]*(1+mue) + mue*Fce[d+2])/(2*mue+1)
        Fce[d + 1] = (Fce[d+2]*(1+mue) + mue*Fce[d-1])/(2*mue+1)
        
        FIP[d] = (FIP[d-1]*(1+muIP) + muIP*FIP[d+2])/(2*muIP+1)
        FIP[d + 1] = (FIP[d+2]*(1+muIP) + muIP*FIP[d-1])/(2*muIP+1)

    return np.concatenate([Fc, Fce, Fh, FIP])


# The right-hand sides of the ODEs
def coscrhs(t, s):
    c = s[: N]  # calcium
    ce = s[N : 2 * N]  # ER calcium
    h = s[2 * N : 3 * N]  # h
    IP = s[3 * N :]  # IP3

    # the equations for the modal model
    phi_c = c ** 4 / (c ** 4 + Kc ** 4)
    phi_p = IP ** 2 / (Kp ** 2 + IP ** 2)
    phi_p_down = Kp ** 2 / (Kp ** 2 + IP ** 2)

    Jpm = Vpm * c ** 2 / (Kpm ** 2 + c ** 2)
    Jin = alpha0 + Vsocc * (Ksocc ** 4 / (Ksocc ** 4 + ce ** 4))
    h_inf = Kh ** 4 / (Kh ** 4 + c ** 4)
    tau = tau_max * Ktau ** 4 / (Ktau ** 4 + c ** 4)
    Jserca = Vserca * (c * c - Kbar * ce * ce) / (c * c + kserca * kserca)

    beta = phi_p * phi_c * h
    alpha = phi_p_down * (1 - phi_c * h_inf)
    Po = beta / (beta + 0.4 * (beta + alpha))
    Jipr = kf * Po * (ce - c)

    dcdt = Jipr - Jserca + delta * (Jin - Jpm)
    dcedt = gamma * (Jserca - Jipr)
    dhdt = (h_inf - h) / tau
    dIPdt = Vplc * c ** 2 / (Kplc ** 2 + c ** 2) - kp * IP

    return np.concatenate([dcdt, dcedt, dhdt, dIPdt])



# Set parameters
tau_max = 100
delta = 1.5
Ktau = 0.1
tauP = 1
Kc = 0.2
Kh = 0.08
kf = 10

Vserca = 0.9
kserca = 0.2
Kbar = 0.00001957
Kp = 0.2

gamma = 5.5
Vpm = 0.11
Kpm = 0.3
alpha0 = 0.0027
Vsocc = 0.07
Ksocc = 8

Vplc = 0.02  # for a regenerating wave
#Vplc = 0    # for no regeneration
Kplc = 0.1
kp = 0.1

Dc = 5  # Ca diffusion coefficient
De = 5  # ER Ca diffusion coefficient
DIP = 200  # IP3 diffusion coefficient
FIP = 10  # intercellular IP3 permeability
Fc = 0  # intercellular Ca permeability
Fe = 0  # intercellular ER Ca permeability

# Geometry parameters
Ncell = 5  # number of cells
N = 500  # number of spatial grid points. Must be divisible by Ncell.
nc = N/Ncell  # number of spatial grid points in each cell.
L = 150  # length of domain
delx = L / N

muc = Fc * delx / Dc  # used in the cell boundary terms
mue = Fe * delx / De
muIP = FIP * delx / DIP

# Set initial data
# all variables start flat except for the IP3 pulse at the left
X = delx * np.arange(1, N + 1)
c0 = 0.08 * np.ones(N)
ce0 = 15 * np.ones(N)
h0 = 0.45 * np.ones(N)
IP0 = np.zeros(N)
IP0[:25] = 5  # IP3 stimulus at left of domain
init = np.concatenate([c0, ce0, h0, IP0])

# Specify the output points
t_end = 20  # total time to run simulation
n_out = 501  # number of output times
tspan = np.linspace(0, t_end, n_out)

# Integrate the system of ODEs
sol = solve_ivp(
    pde_rhs,
    t_span=(tspan[0], tspan[-1]),
    y0=init,
    method="BDF",
    t_eval=tspan,
    jac=None,
    rtol=1e-6,
    atol=1e-8
)

c = sol.y[:N]
ce = sol.y[N : 2 * N]
IP = sol.y[3 * N :]

# Plot a movie
plt.figure(1)
for i in range(n_out):
    plt.plot(X, c[:, i], linewidth=2)
    plt.ylim([0, 1])
    plt.xlim([0, L])
    plt.xlabel("x", fontsize=20)
    plt.ylabel(r"$[Ca^{2+}] (\mu {\rm M})$", fontsize=20)
    plt.pause(0.1)


plt.figure(2)
plt.imshow(np.transpose(c), aspect='auto', extent=[0, L, t_end, 0], cmap='viridis')
plt.colorbar(label=r'$Ca^{2+}$ concentration')
plt.xlabel("x ($\mu m$)")
plt.ylabel("time (sec)")
plt.show()

File: chap07_calcium/intercell_cal_wave.m

%  -------------------------------------------------------------------
%
%   Use the method of lines to compute an intercellular calcium wave. This
%   version uses the modal model of the IP3 receptor, but this can be
%   changed. Here, IP3 is also diffusing and has simple reaction terms.
%
%   There are jump conditions specifying
%   the flux across each intercellular boundary.
%
%   For Chapter 7, Section 7.10.2 of
%   Keener and Sneyd, Mathematical Physiology, 3rd Edition, Springer.
% 
%   Written by James Keener and James Sneyd.
% 
%  ------------------------------------------------------------------- 

clear all
close all
clc
set(0,                           ...
'defaultaxesfontsize', 20,   ...
'defaultaxeslinewidth', 2.0, ...
'defaultlinelinewidth', 2.0);

global p

% parameters

p.tau_max=100;
p.delta=1.5;
p.Ktau=0.1;
p.tauP=1;
p.Kc=0.2;
p.Kh=0.08;
p.kf=10;
 
p.Vserca=0.9;
p.kserca=0.2;
p.Kbar=0.00001957;
p.Kp=0.2;

p.gamma=5.5;
p.Vpm=0.11;
p.Kpm=0.3;
p.alpha0=.0027;
p.Vsocc = .07;
p.Ksocc=8;

p.Vplc = 0.02; % for a regenerating wave
%p.Vplc = 0;     % for no regeneration
p.Kplc = 0.1;
p.kp = 0.1;

p.Dc = 5;       % Ca diffusion coefficient
p.De = 5;       % ER Ca diffusion coefficient
p.DIP = 200;    % IP3 diffusion coefficient
p.FIP = 10;     % intercellular IP3 permeability
p.Fc  = 0;      % intercellular Ca permeability
p.Fe = 0;       % intercellular ER Ca permeability


% Geometry parameters
p.Ncell = 5;        % number of cells
p.N = 500;          % number of spatial grid points. Must be divisible by p.Ncell.
p.nc = p.N/p.Ncell; % number of spatial grid points in each cell.
p.L = 150;          % length of domain
p.delx = p.L/p.N;

p.muc = p.Fc*p.delx/p.Dc;  % used in the cell boundary terms
p.mue = p.Fe*p.delx/p.De;
p.muIP = p.FIP*p.delx/p.DIP;

% set initial data
% all variables start flat except for the IP3 pulse at the left
X = p.delx*(1:p.N)';
c0 = 0.08*ones(p.N,1);
ce0 = 15*ones(p.N,1);
h0 = 0.45*ones(p.N,1);
IP0 = zeros(p.N,1);  
IP0(1:25) = 5;              % IP3 stimulus at left of domain
init = [c0;ce0;h0;IP0];

% specify the output points
t_end = 20;     % total time to run simulation
n_out = 501;    % number of output times
tspan = linspace(0,t_end,n_out);

%%%%%%%%

% integrate the system of odes. It's a DAE because of the jump conditions.
% The zeros of the mass matrix, M, tells you where these jumps occur.

% set up the mass matrix. It needs two zero entries on either side of each
% intecellular boundary, but only for the diffusing species (so not h).
M = eye(4*p.N);
for j = 1:p.Ncell-1
    d = j*p.nc;
    M(d,d) = 0;
    M(d+1,d+1) = 0;
    M(p.N+d,p.N+d) = 0;
    M(p.N+d+1,p.N+d+1) = 0;
    M(3*p.N+d,3*p.N+d) = 0;
    M(3*p.N+d+1,3*p.N+d+1) = 0;
end

options = odeset('Mass',M);
[T,S] = ode15s( @(t,x)pdeRHS(t,x),tspan,init,options); 
c = S(:,1:p.N);             % calcium
ce = S(:,p.N+1:2*p.N);      % ER calcium
IP = S(:,3*p.N+1:4*p.N);    % IP3

% plot a movie  
figure(1)
for i=1:n_out
    plot(X,c(i,:),'linewidth',2)
    ylim([0,1])
    xlim([0,p.L])
    pause(0.1)
end
xlabel('x','fontsize',20)
ylabel('Ca^{++}(x,t)','fontsize',20)

figure(2)
imagesc(X,T,c)
colorbar
xlabel('x (\mu m)')
ylabel('t (s)')





%% the right hand side for pde (MoL) simulation:

function s_prime = pdeRHS(t,s)
global p
 
% There are four variables:  c, ce, h, IP
% only c, ce and IP are diffusing

c = s(1:p.N); % calcium
ce = s(p.N+1:2*p.N); % ER calcium
h = s(2*p.N+1:3*p.N); % h
IP = s(3*p.N+1:4*p.N); % IP3

lam_c = p.Dc/p.delx^2;
lam_e = p.De/p.delx^2;
lam_p = p.DIP/p.delx^2;

out=coscrhs(t,s);
sc = [1;2*ones(p.N-2,1);1];

Fc  = lam_c*(-sc.*c + [0;c(1:end-1)]+[c(2:end);0]) + out(1:p.N);
Fce = lam_e*(-sc.*ce + [0;ce(1:end-1)]+[ce(2:end);0])+ out(p.N+1:2*p.N);           
Fh  = out(2*p.N+1:3*p.N);                                                   % reaction only, no diffusion
FIP  = lam_p*(-sc.*IP +[ 0;IP(1:end-1)]+[IP(2:end);0]) + out(3*p.N+1:4*p.N) ;

% now put in the intercellular boundaries. A zero term on each side of the
% boundary.
for j=1:p.Ncell-1
    d = j*p.nc;

    Fc(d)   = c(d) - c(d-1) + p.muc*(c(d) - c(d+1));
    Fc(d+1) = c(d+2) - c(d+1) + p.muc*(c(d) - c(d+1));
    
    Fce(d)   = ce(d) - ce(d-1) + p.mue*(ce(d) - ce(d+1));
    Fce(d+1) = ce(d+2) - ce(d+1) + p.mue*(ce(d) - ce(d+1));
    
    FIP(d)   = IP(d) - IP(d-1) + p.muIP*(IP(d) - IP(d+1));
    FIP(d+1) = IP(d+2) - IP(d+1) + p.muIP*(IP(d) - IP(d+1));
end

s_prime = [Fc;Fce;Fh;FIP];

end
 

%% the right hand sides of the odes

function out = coscrhs(t,s)
global p

c = s(1:p.N); % calcium
ce = s(p.N+1:2*p.N); % ER calcium
h = s(2*p.N+1:3*p.N); % h
IP = s(3*p.N+1:4*p.N); % IP3

% the equations for the modal model
phi_c=c.^4./(c.^4+p.Kc^4);
phi_p=IP.^2./(p.Kp^2+IP.^2);
phi_p_down=p.Kp^2./(p.Kp^2+IP.^2);

Jpm = p.Vpm*c.^2./(p.Kpm^2+c.^2);
Jin = p.alpha0 + p.Vsocc*(p.Ksocc^4./(p.Ksocc^4+ce.^4));
h_inf=p.Kh^4./(p.Kh^4+c.^4);
tau = p.tau_max*p.Ktau^4./(p.Ktau^4+c.^4);
Jserca=p.Vserca*(c.*c-p.Kbar*ce.*ce)./(c.*c+p.kserca*p.kserca);

beta = phi_p.*phi_c.*h;
alpha = phi_p_down.*(1-phi_c.*h_inf);
Po=beta./(beta + 0.4*(beta+alpha));
Jipr = p.kf*Po.*(ce-c);

dcdt = Jipr-Jserca+p.delta*(Jin-Jpm);
dcedt = p.gamma*(Jserca-Jipr);
dhdt = (h_inf-h)./tau;
dIPdt = p.Vplc*c.^2./(p.Kplc^2 + c.^2) - p.kp*IP;                          
out = [dcdt;dcedt;dhdt;dIPdt];

end
 
 